File: libs/power-log-parser/src/lib/node-parser.ts
import { Node } from './models';
import { GameEventProvider } from './game-event';
import { EventQueueHandler } from './event-queue-handler';
import { ControlsManager } from './controls/controls-manager';
import { Logger } from './logger';
import type { ActionParser } from './action-parser';
import type { StateFacade } from './state/state-facade';
import type { ParserState } from './state/parser-state';
import { StateType, INodeParser } from './state/parser-state';
import {
	ArmorChangeParser,
	AttackOnBoardParser,
	AttackParser,
	BallerBuffChangedParser,
	BattlegroundFreezeParser,
	BattlegroundsActivePlayerBoardParser,
	BattlegroundsBattleOverParser,
	BattlegroundsBuddyGainedParser,
	BattlegroundsDuoTeammatePlayerBoardParser,
	BattlegroundsHeroKilledParser,
	BattlegroundsHeroRerollParser,
	BattlegroundsHeroSelectedParser,
	BattlegroundsHeroSelectionParser,
	BattlegroundsMinionsBoughtParser,
	BattlegroundsMinionsSoldParser,
	BattlegroundsNextOpponentParser,
	BattlegroundsOpponentRevealedParser,
	BattlegroundsPlayerBoardParser,
	BattlegroundsPlayerLeaderboardPlaceUpdatedParser,
	BattlegroundsPlayerTechLevelUpdatedParser,
	BattlegroundsQuestCompletedParser,
	BattlegroundsQuestRevealedParser,
	BattlegroundsQuestRewardDestroyedParser,
	BattlegroundsQuestRewardEquippedParser,
	BattlegroundsRerollParser,
	BattlegroundsRewardGainedParser,
	BattlegroundsTavernPrizesParser,
	BattlegroundsTimewarpedTavernActiveParser,
	BattlegroundsTrinketSelectedParser,
	BattlegroundsTrinketSelectionParser,
	BattlegroundsTriplesCountUpdatedParser,
	BeetleArmyChangedParser,
	BloodGemBuffChangedParser,
	BurnedCardParser,
	CardBackToDeckParser,
	CardBuffedInHandParser,
	CardChangedParser,
	CardDrawFromDeckParser,
	CardForgedParser,
	CardPlayedFromEffectParser,
	CardPlayedFromHandParser,
	CardPresentOnGameStartParser,
	CardRemovedFromBoardParser,
	CardRemovedFromDeckParser,
	CardRemovedFromHandParser,
	CardRevealedParser,
	CardStolenParser,
	CardTradedParser,
	CardUpdatedInDeckParser,
	CardsShuffledIntoDeckParser,
	ChangeCardCreatorParser,
	ChoosingOptionsParser,
	ConstructedAnomalyParser,
	CopiedFromEntityIdParser,
	CorpsesChangedParser,
	CorpsesSpentThisGameParser,
	CostChangedParser,
	CounterTriggerParser,
	CounterWillTriggerParser,
	CreateCardInDeckParser,
	CreateCardInGraveyardParser,
	CthunParser,
	DamageParser,
	DataScriptChangedParser,
	DeathrattleTriggeredParser,
	DecklistUpdateParser,
	DiscardedCardParser,
	DungeonRunStepParser,
	EndOfEchoInHandParser,
	EntityChosenParser,
	EntityUpdateParser,
	ExcavateTierChangedParser,
	FatigueParser,
	FirstPlayerParser,
	FullEntityParser,
	GalakrondInvokedParser,
	GameCleanupParser,
	GameEndParser,
	GameResetParser,
	GameRunningParser,
	HealingParser,
	HealthDefChangeParser,
	HeraldColossalAmountParser,
	HeroChangedParser,
	HeroEnchantmentAttachedParser,
	HeroEnchantmentDetachedParser,
	HeroPowerChangedParser,
	HeroPowerUsedParser,
	HeroRevealedParser,
	HideEntityParser,
	ImmolateChangedParser,
	InitialCardInDeckParser,
	JadeGolemParser,
	LinkedEntityParser,
	LocalPlayerLeaderboardPlaceChangedParser,
	LocationDestroyedParser,
	LocationUsedParser,
	MainStepReadyParser,
	MaxResourcesChangedParser,
	MercenariesAbilityActivatedParser,
	MercenariesAbilityCooldownUpdatedParser,
	MercenariesAbilityRevealedParser,
	MercenariesHeroRevealedParser,
	MercenariesQueuedAbilityParser,
	MercenariesSelectedTargetParser,
	MindrenderIlluciaParser,
	MinionBackOnBoardParser,
	MinionDiedParser,
	MinionGoDormantParser,
	MinionOnBoardAttackUpdatedParser,
	MinionSummonedParser,
	MinionsWillDieParser,
	MonsterRunStepParser,
	MulliganDealingParser,
	MulliganDoneParser,
	MulliganInputParser,
	NewGameParser,
	NumCardsDrawnThisTurnParser,
	NumCardsPlayedThisTurnParser,
	OverloadedCrystalsParser,
	ParentCardChangedParser,
	PassiveBuffParser,
	QuestCompletedParser,
	ReceiveCardInHandParser,
	RecruitParser,
	RemovedFromHistoryParser,
	ResourcesUsedThisTurnParser,
	RumbleRunStepParser,
	SecretCreatedInGameParser,
	SecretDestroyedParser,
	SecretPlayedFromDeckParser,
	SecretPlayedFromHandParser,
	SecretTriggeredParser,
	SecretWillTriggeredParser,
	ShowEntityParser,
	ShuffleDeckParser,
	SpecialCardPowerParser,
	SpecialTargetParser,
	SpawnTimeCountChangedParser,
	StarshipLaunchedParser,
	StartOfGameTriggerParser,
	TotalMagnetizeChangedParser,
	TouristRevealedParser,
	TurnCleanupParser,
	TurnDurationUpdateParser,
	TurnStartParser,
	WeaponDestroyedParser,
	WeaponEquippedParser,
	WheelOfDeathCounterParser,
	WhizbangDeckParser,
	WinnerParser,
	ZoneChangeParser,
	ZonePositionChangedParser,
} from './parsers';

export class NodeParser implements INodeParser {
	private QueueHandler: EventQueueHandler;
	private StateFacade: StateFacade;
	private ParserState!: ParserState;
	private StateType: StateType;
	private Controller: ControlsManager;

	private _parsers: ActionParser[] | null = null;
	private _unfilteredParsers: ActionParser[] | null = null;
	private get parsers(): ActionParser[] {
		if (this._parsers != null) {
			return this._parsers;
		}
		if (this.ParserState == null) {
			return [];
		}

		if (this._unfilteredParsers == null) {
			this._unfilteredParsers = this.BuildActionParsers(this.ParserState, this.StateType);
		}

		if (
			this.StateFacade?.GsState?.CurrentGame?.GameType == null ||
			this.StateFacade?.GsState?.CurrentGame?.GameType === -1
		) {
			return this._unfilteredParsers;
		}

		this._parsers = this._unfilteredParsers.filter((p) => this.Controller.Applies(p));
		return this._parsers;
	}

	constructor(queueHandler: EventQueueHandler, stateFacade: StateFacade, stateType: StateType) {
		this.QueueHandler = queueHandler;
		this.StateFacade = stateFacade;
		this.StateType = stateType;
		this.Controller = new ControlsManager(stateFacade, stateType);
	}

	Reset(parserState: ParserState, helper: StateFacade): void {
		this.StateFacade = helper;
		this.ParserState = parserState;
		this.QueueHandler.Reset(helper);
		this._parsers = null;
		this._unfilteredParsers = null;
	}

	NewNode(node: Node, stateType: StateType): void {
		if (node == null) {
			return;
		}
		for (const parser of this.parsers) {
			if (parser.AppliesOnNewNode(node, stateType)) {
				try {
					const providers = parser.CreateGameEventProviderFromNew(node);
					if (providers != null) {
						this.EnqueueGameEvent(providers);
					}
				} catch (e: any) {
					Logger.Log('ERROR: Exception while parsing node', e?.message ?? e);
					Logger.Log(node.CreationLogLine ?? '', '');
					Logger.Log(e?.stack ?? '', '');
				}
			}
		}
	}

	CloseNode(node: Node, stateType: StateType): void {
		if (node == null) {
			return;
		}
		for (const parser of this.parsers) {
			if (!node.Closed && parser.AppliesOnCloseNode(node, stateType)) {
				try {
					const providers = parser.CreateGameEventProviderFromClose(node);
					if (providers != null && providers.length > 0) {
						this.EnqueueGameEvent(providers);
					}
				} catch (e: any) {
					Logger.Log('ERROR: Exception while parsing node', e?.message ?? e);
					Logger.Log(node.CreationLogLine ?? '', '');
					Logger.Log(e?.stack ?? '', '');
				}
			}
		}
		node.Closed = true;
	}

	EnqueueGameEvent(providers: GameEventProvider[]): void {
		this.QueueHandler.EnqueueGameEvent(providers);
	}

	ClearQueue(): void {
		this.QueueHandler.ClearQueue();
	}

	private BuildActionParsers(parserState: ParserState, stateType: StateType): ActionParser[] {
		if (stateType === StateType.GameState) {
			return [
				new NewGameParser(parserState, this.StateFacade),
				new TurnCleanupParser(parserState, this.StateFacade),
				new GameCleanupParser(parserState, this.StateFacade),
				new SecretWillTriggeredParser(parserState, this.StateFacade),
				new CounterWillTriggerParser(parserState, this.StateFacade),
				new MinionsWillDieParser(parserState, this.StateFacade),
				new ChoosingOptionsParser(parserState, this.StateFacade),
				new EntityChosenParser(parserState, this.StateFacade),
				new BattlegroundsHeroSelectedParser(parserState, this.StateFacade),
				new BattlegroundsDuoTeammatePlayerBoardParser(parserState, this.StateFacade),
				new BattlegroundsActivePlayerBoardParser(parserState, this.StateFacade),
				new BattlegroundsPlayerBoardParser(parserState, this.StateFacade),
			];
		}
		return [
			new GameResetParser(parserState, this.StateFacade),

			new HideEntityParser(parserState, this.StateFacade),
			new ShowEntityParser(parserState, this.StateFacade),
			new FullEntityParser(parserState, this.StateFacade),

			new MinionSummonedParser(parserState, this.StateFacade),

			new CardRevealedParser(parserState, this.StateFacade),

			new MercenariesHeroRevealedParser(parserState, this.StateFacade),
			new MercenariesAbilityRevealedParser(parserState, this.StateFacade),
			new MercenariesAbilityActivatedParser(parserState, this.StateFacade),
			new MercenariesAbilityCooldownUpdatedParser(parserState, this.StateFacade),
			new MercenariesQueuedAbilityParser(parserState, this.StateFacade),
			new MercenariesSelectedTargetParser(parserState, this.StateFacade),

			new BattlegroundsPlayerBoardParser(parserState, this.StateFacade),
			new ShuffleDeckParser(parserState, this.StateFacade),

			new WinnerParser(parserState, this.StateFacade),
			new GameEndParser(parserState, this.StateFacade),
			new TurnStartParser(parserState, this.StateFacade),
			new FirstPlayerParser(parserState, this.StateFacade),
			new MainStepReadyParser(parserState, this.StateFacade),
			new CardPlayedFromHandParser(parserState, this.StateFacade),
			new CardPlayedFromEffectParser(parserState, this.StateFacade),
			new SecretPlayedFromHandParser(parserState, this.StateFacade),
			new MulliganInputParser(parserState),
			new MulliganDealingParser(parserState),
			new MulliganDoneParser(parserState),
			new RumbleRunStepParser(parserState, this.StateFacade),
			new DungeonRunStepParser(parserState, this.StateFacade),
			new MonsterRunStepParser(parserState, this.StateFacade),
			new PassiveBuffParser(parserState, this.StateFacade),
			new CardPresentOnGameStartParser(parserState, this.StateFacade),
			new CardDrawFromDeckParser(parserState, this.StateFacade),
			new ReceiveCardInHandParser(parserState, this.StateFacade),
			new CardBackToDeckParser(parserState, this.StateFacade),
			new CardTradedParser(parserState, this.StateFacade),
			new DiscardedCardParser(parserState, this.StateFacade),
			new CardRemovedFromDeckParser(parserState, this.StateFacade),
			new CreateCardInDeckParser(parserState, this.StateFacade),
			new EndOfEchoInHandParser(parserState, this.StateFacade),
			new CardChangedParser(parserState, this.StateFacade),
			new CardUpdatedInDeckParser(parserState, this.StateFacade),
			new CardRemovedFromHandParser(parserState, this.StateFacade),
			new CardRemovedFromBoardParser(parserState, this.StateFacade),
			new MinionOnBoardAttackUpdatedParser(parserState, this.StateFacade),
			new RecruitParser(parserState, this.StateFacade),
			new MinionBackOnBoardParser(parserState, this.StateFacade),
			new HeroRevealedParser(parserState, this.StateFacade),
			new InitialCardInDeckParser(parserState, this.StateFacade),
			new FatigueParser(parserState, this.StateFacade),
			new DamageParser(parserState, this.StateFacade),
			new HealingParser(parserState, this.StateFacade),
			new BurnedCardParser(parserState, this.StateFacade),
			new RemovedFromHistoryParser(parserState, this.StateFacade),
			new MinionDiedParser(parserState, this.StateFacade),
			new SecretPlayedFromDeckParser(parserState, this.StateFacade),
			new SecretCreatedInGameParser(parserState, this.StateFacade),
			new SecretDestroyedParser(parserState, this.StateFacade),
			new ArmorChangeParser(parserState, this.StateFacade),
			new CorpsesChangedParser(parserState, this.StateFacade),
			new HeraldColossalAmountParser(parserState, this.StateFacade),
			new MaxResourcesChangedParser(parserState, this.StateFacade),
			new ExcavateTierChangedParser(parserState, this.StateFacade),
			new CardStolenParser(parserState, this.StateFacade),
			new SecretTriggeredParser(parserState, this.StateFacade),
			new CounterTriggerParser(parserState, this.StateFacade),
			new QuestCompletedParser(parserState, this.StateFacade),
			new DeathrattleTriggeredParser(parserState, this.StateFacade),
			new HealthDefChangeParser(parserState, this.StateFacade),
			new ChangeCardCreatorParser(parserState, this.StateFacade),
			new LocalPlayerLeaderboardPlaceChangedParser(parserState, this.StateFacade),
			new HeroPowerChangedParser(parserState, this.StateFacade),
			new HeroChangedParser(parserState, this.StateFacade),
			new WeaponEquippedParser(parserState, this.StateFacade),
			new WeaponDestroyedParser(parserState, this.StateFacade),
			new LocationDestroyedParser(parserState, this.StateFacade),
			new HeroEnchantmentAttachedParser(parserState, this.StateFacade),
			new HeroEnchantmentDetachedParser(parserState, this.StateFacade),
			new CardsShuffledIntoDeckParser(parserState, this.StateFacade),

			new BattlegroundsPlayerTechLevelUpdatedParser(parserState, this.StateFacade),
			new BattlegroundsBuddyGainedParser(parserState, this.StateFacade),
			new BattlegroundsQuestRevealedParser(parserState, this.StateFacade),
			new BattlegroundsQuestCompletedParser(parserState, this.StateFacade),
			new BattlegroundsRewardGainedParser(parserState, this.StateFacade),
			new BattlegroundsPlayerLeaderboardPlaceUpdatedParser(parserState, this.StateFacade),
			new BattlegroundsHeroSelectionParser(parserState, this.StateFacade),
			new BattlegroundsHeroRerollParser(parserState, this.StateFacade),
			new BattlegroundsNextOpponentParser(parserState),
			new BattlegroundsTriplesCountUpdatedParser(parserState, this.StateFacade),
			new BattlegroundsOpponentRevealedParser(parserState, this.StateFacade),
			new BattlegroundsHeroSelectedParser(parserState, this.StateFacade),
			new BattlegroundsBattleOverParser(parserState, this.StateFacade),
			new BattlegroundsRerollParser(parserState, this.StateFacade),
			new BattlegroundFreezeParser(parserState, this.StateFacade),
			new BattlegroundsMinionsBoughtParser(parserState, this.StateFacade),
			new BattlegroundsMinionsSoldParser(parserState, this.StateFacade),
			new BattlegroundsHeroKilledParser(parserState, this.StateFacade),
			new BattlegroundsQuestRewardEquippedParser(parserState, this.StateFacade),
			new BattlegroundsQuestRewardDestroyedParser(parserState, this.StateFacade),
			new BloodGemBuffChangedParser(parserState, this.StateFacade),
			new BeetleArmyChangedParser(parserState, this.StateFacade),
			new BallerBuffChangedParser(parserState, this.StateFacade),
			new TotalMagnetizeChangedParser(parserState, this.StateFacade),
			new BattlegroundsTrinketSelectedParser(parserState, this.StateFacade),
			new BattlegroundsTrinketSelectionParser(parserState, this.StateFacade),
			new BattlegroundsTimewarpedTavernActiveParser(parserState, this.StateFacade),

			new DecklistUpdateParser(parserState, this.StateFacade),
			new GameRunningParser(parserState, this.StateFacade),
			new AttackParser(parserState, this.StateFacade),
			new NumCardsPlayedThisTurnParser(parserState, this.StateFacade),
			new NumCardsDrawnThisTurnParser(parserState, this.StateFacade),
			new HeroPowerUsedParser(parserState, this.StateFacade),
			new GalakrondInvokedParser(parserState, this.StateFacade),
			new CardBuffedInHandParser(parserState, this.StateFacade),
			new MinionGoDormantParser(parserState, this.StateFacade),
			new AttackOnBoardParser(parserState, this.StateFacade),
			new JadeGolemParser(parserState, this.StateFacade),
			new CthunParser(parserState, this.StateFacade),
			new EntityUpdateParser(parserState, this.StateFacade),
			new ResourcesUsedThisTurnParser(parserState, this.StateFacade),
			new WhizbangDeckParser(parserState, this.StateFacade),
			new BattlegroundsTavernPrizesParser(parserState),
			new LinkedEntityParser(parserState, this.StateFacade),
			new ZoneChangeParser(parserState, this.StateFacade),
			new ZonePositionChangedParser(parserState, this.StateFacade),
			new CostChangedParser(parserState, this.StateFacade),
			new SpawnTimeCountChangedParser(parserState, this.StateFacade),
			new TurnDurationUpdateParser(parserState, this.StateFacade),
			new StartOfGameTriggerParser(parserState, this.StateFacade),
			new DataScriptChangedParser(parserState, this.StateFacade),
			new ImmolateChangedParser(parserState, this.StateFacade),
			new OverloadedCrystalsParser(parserState, this.StateFacade),
			new CorpsesSpentThisGameParser(parserState, this.StateFacade),
			new ConstructedAnomalyParser(parserState, this.StateFacade),
			new TouristRevealedParser(parserState, this.StateFacade),
			new ParentCardChangedParser(parserState, this.StateFacade),
			new StarshipLaunchedParser(parserState, this.StateFacade),

			new CardForgedParser(parserState, this.StateFacade),
			new LocationUsedParser(parserState, this.StateFacade),
			new CreateCardInGraveyardParser(parserState, this.StateFacade),
			new MindrenderIlluciaParser(parserState, this.StateFacade),
			new SpecialCardPowerParser(parserState, this.StateFacade),
			new WheelOfDeathCounterParser(parserState, this.StateFacade),

			new CopiedFromEntityIdParser(parserState, this.StateFacade),
			new SpecialTargetParser(parserState, this.StateFacade),
		];
	}
}

File: tools/linux-helpers/hs-converter/HearthstoneReplays/Events/NodeParser.cs
#region
using System;
using System.Timers;
using System.Collections.Generic;
using System.Linq;
using System.Threading.Tasks;
using HearthstoneReplays.Enums;
using HearthstoneReplays.Parser.ReplayData.Entities;
using HearthstoneReplays.Parser.ReplayData.GameActions;
using HearthstoneReplays.Parser;
using HearthstoneReplays.Events.Parsers;
using HearthstoneReplays.Events.Parsers.Controls;
#endregion

namespace HearthstoneReplays.Events
{
    public class NodeParser
    {
        private EventQueueHandler QueueHandler { get; set; }
        private StateFacade StateFacade { get; set; }
        private ParserState ParserState { get; set; }
        private StateType StateType { get; set; }

        private ControlsManager Controller { get; set; }

        private List<ActionParser> _parsers;
        private List<ActionParser> parsers { 
            get {
                if (_parsers != null)
                {
                    return _parsers;
                }
                if (ParserState == null)
                {
                    return new List<ActionParser>();
                }
                if (StateFacade?.GsState?.CurrentGame?.GameType == null || ParserState == null || StateFacade?.GsState?.CurrentGame?.GameType == -1)
                {
                    // For NEW_GAME event
                    return BuildActionParsers(ParserState, StateType);
                }
                var allParsers = BuildActionParsers(ParserState, StateType);
                this._parsers = allParsers
                    .Where(p => Controller.Applies(p))
                    .ToList();
                return this._parsers;
            } 
        }

        // Feed it the PTL parser, as it's the latest one to get the meta data
        public NodeParser(EventQueueHandler queueHandler, StateFacade stateFacade, StateType stateType)
        {
            QueueHandler = queueHandler;
            StateFacade = stateFacade;
            StateType = stateType;
            Controller = new ControlsManager(stateFacade, stateType);
        }

        public void Reset(ParserState ParserState, StateFacade helper)
        {
            this.StateFacade = helper;
            this.ParserState = ParserState;
            this.QueueHandler.Reset(StateFacade);
            this._parsers = null;
        }

        public void NewNode(Node node, StateType stateType)
        {
            if (node == null)
            {
                return;
            }
            foreach (ActionParser parser in parsers)
            {
                if (parser.AppliesOnNewNode(node, stateType))
                {
                    try
                    {
                        List<GameEventProvider> providers = parser.CreateGameEventProviderFromNew(node);
                        if (providers != null)
                        {
                            EnqueueGameEvent(providers);
                        }
                    }
                    catch (Exception e)
                    {
                        Logger.Log("ERROR: Exception while parsing node", e.Message);
                        Logger.Log(node.CreationLogLine, "");
                        Logger.Log(e.StackTrace.ToString(), "");
                    }
                }
            }
        }

        public void CloseNode(Node node, StateType stateType)
        {
            if (node == null)
            {
                return;
            }
            foreach (ActionParser parser in parsers)
            {
                if (!node.Closed && parser.AppliesOnCloseNode(node, stateType))
                {
                    try
                    {
                        List<GameEventProvider> providers = parser.CreateGameEventProviderFromClose(node);
                        if (providers != null && providers.Count > 0)
                        {
                            EnqueueGameEvent(providers);
                        }
                    }
                    catch (Exception e)
                    {
                        Logger.Log("ERROR: Exception while parsing node", e.Message);
                        Logger.Log(node.CreationLogLine, "");
                        Logger.Log(e.StackTrace.ToString(), "");
                    }
                }
            }
            // Make sure we don't process the same node twice
            node.Closed = true;
        }

        public void EnqueueGameEvent(List<GameEventProvider> providers)
        {
            QueueHandler.EnqueueGameEvent(providers);
        }

        public void ClearQueue()
        {
            QueueHandler.ClearQueue();
        }

        private List<ActionParser> BuildActionParsers(ParserState ParserState, StateType stateType)
        {
            if (stateType == StateType.GameState)
            {
                return new List<ActionParser>()
                {
                    new NewGameParser(ParserState, StateFacade),
                    new TurnCleanupParser(ParserState, StateFacade),
                    new GameCleanupParser(ParserState, StateFacade),
                    new SecretWillTriggeredParser(ParserState, StateFacade),
                    new CounterWillTriggerParser(ParserState, StateFacade),
                    new MinionsWillDieParser(ParserState, StateFacade),
                    new ChoosingOptionsParser(ParserState, StateFacade),
                    // Will parse the GameState logs (because choices are logged only in GS), but uses the PTL states
                    new EntityChosenParser(ParserState, StateFacade),

                    new BattlegroundsHeroSelectedParser(ParserState, StateFacade),
                    new BattlegroundsDuoTeammatePlayerBoardParser(ParserState, StateFacade),
                    //new BattlegroundsBattleStartingParser(ParserState, StateFacade),
                    new BattlegroundsActivePlayerBoardParser(ParserState, StateFacade),
                    new BattlegroundsPlayerBoardParser(ParserState, StateFacade),
                };
            }
            else
            {
                return new List<ActionParser>()
                {
                    new GameResetParser(ParserState, StateFacade),

                    // Ordering is important,
                    new HideEntityParser(ParserState, StateFacade),
                    new ShowEntityParser(ParserState, StateFacade),
                    new FullEntityParser(ParserState, StateFacade),

                    // we want to have "ability revealed" before 
                    // "ability updated" (done by the EntityUpdateParser). Same with CardRevealed
                    // Also, MINION_SUMMONED need to happen before any Equipment / Ability revealed
                    new MinionSummonedParser(ParserState, StateFacade),

                    new CardRevealedParser(ParserState, StateFacade),

                    new MercenariesHeroRevealed(ParserState, StateFacade),
                    new MercenariesAbilityRevealedParser(ParserState, StateFacade),
                    new MercenariesAbilityActivatedParser(ParserState, StateFacade),
                    new MercenariesAbilityCooldownUpdatedParser(ParserState, StateFacade),
                    new MercenariesQueuedAbilityParser(ParserState, StateFacade),
                    new MercenariesSelectedTargetParser(ParserState, StateFacade),

                    new BattlegroundsPlayerBoardParser(ParserState, StateFacade),
                    new ShuffleDeckParser(ParserState, StateFacade),

                    new WinnerParser(ParserState, StateFacade),
                    new GameEndParser(ParserState, StateFacade),
                    new TurnStartParser(ParserState, StateFacade),
                    new FirstPlayerParser(ParserState, StateFacade),
                    new MainStepReadyParser(ParserState),
                    new CardPlayedFromHandParser(ParserState, StateFacade),
                    new CardPlayedFromEffectParser(ParserState, StateFacade),
                    new SecretPlayedFromHandParser(ParserState, StateFacade),
                    new MulliganInputParser(ParserState),
                    new MulliganDealingParser(ParserState),
                    new MulliganDoneParser(ParserState),
                    new RumbleRunStepParser(ParserState, StateFacade),
                    new DungeonRunStepParser(ParserState, StateFacade),
                    new MonsterRunStepParser(ParserState, StateFacade),
                    new PassiveBuffParser(ParserState, StateFacade),
                    new CardPresentOnGameStartParser(ParserState, StateFacade),
                    new CardDrawFromDeckParser(ParserState, StateFacade),
                    new ReceiveCardInHandParser(ParserState, StateFacade),
                    new CardBackToDeckParser(ParserState, StateFacade),
                    new CardTradedParser(ParserState, StateFacade),
                    new DiscardedCardParser(ParserState, StateFacade),
                    new CardRemovedFromDeckParser(ParserState, StateFacade),
                    new CreateCardInDeckParser(ParserState, StateFacade),
                    new EndOfEchoInHandParser(ParserState, StateFacade),
                    new CardChangedParser(ParserState, StateFacade),
                    new CardUpdatedInDeckParser(ParserState, StateFacade),
                    new CardRemovedFromHandParser(ParserState, StateFacade),
                    new CardRemovedFromBoardParser(ParserState, StateFacade),
                    new MinionOnBoardAttackUpdatedParser(ParserState, StateFacade),
                    new RecruitParser(ParserState, StateFacade),
                    new MinionBackOnBoardParser(ParserState, StateFacade),
                    new HeroRevealedParser(ParserState, StateFacade),
                    new InitialCardInDeckParser(ParserState, StateFacade),
                    new FatigueParser(ParserState, StateFacade),
                    new DamageParser(ParserState, StateFacade),
                    new HealingParser(ParserState, StateFacade),
                    new BurnedCardParser(ParserState, StateFacade),
                    new RemovedFromHistoryParser(ParserState, StateFacade),
                    new MinionDiedParser(ParserState, StateFacade),
                    new SecretPlayedFromDeckParser(ParserState, StateFacade),
                    new SecretCreatedInGameParser(ParserState, StateFacade),
                    new SecretDestroyedParser(ParserState, StateFacade),
                    new ArmorChangeParser(ParserState, StateFacade),
                    new CorpsesChangedParser(ParserState, StateFacade),
                    new HeraldColossalAmountParser(ParserState, StateFacade),
                    new MaxResourcesChangedParser(ParserState, StateFacade),
                    new ExcavateTierChangedParser(ParserState, StateFacade),
                    new CardStolenParser(ParserState, StateFacade),
                    new SecretTriggeredParser(ParserState, StateFacade),
                    new CounterTriggerParser(ParserState, StateFacade),
                    new QuestCompletedParser(ParserState, StateFacade),
                    new DeathrattleTriggeredParser(ParserState, StateFacade),
                    new HealthDefChangeParser(ParserState, StateFacade),
                    new ChangeCardCreatorParser(ParserState, StateFacade),
                    new LocalPlayerLeaderboardPlaceChangedParser(ParserState, StateFacade),
                    new HeroPowerChangedParser(ParserState, StateFacade),
                    new HeroChangedParser(ParserState, StateFacade),
                    new WeaponEquippedParser(ParserState, StateFacade),
                    new WeaponDestroyedParser(ParserState, StateFacade),
                    new LocationDestroyedParser(ParserState, StateFacade),
                    new HeroEnchantmentAttachedParser(ParserState, StateFacade),
                    new HeroEnchantmentDetachedParser(ParserState, StateFacade),
                    new CardsShuffledIntoDeckParser(ParserState, StateFacade),

                    new BattlegroundsPlayerTechLevelUpdatedParser(ParserState, StateFacade),
                    new BattlegroundsBuddyGainedParser(ParserState, StateFacade),
                    new BattlegroundsQuestRevealedParser(ParserState, StateFacade),
                    new BattlegroundsQuestCompletedParser(ParserState, StateFacade),
                    new BattlegroundsRewardGainedParser(ParserState, StateFacade),
                    new BattlegroundsPlayerLeaderboardPlaceUpdatedParser(ParserState, StateFacade),
                    new BattlegroundsHeroSelectionParser(ParserState, StateFacade),
                    new BattlegroundsHeroRerollParser(ParserState, StateFacade),
                    new BattlegroundsNextOpponnentParser(ParserState),
                    new BattlegroundsTriplesCountUpdatedParser(ParserState, StateFacade),
                    //new BattlegroundsStartOfCombatParser(ParserState),
                    //new BattlegroundsAfterPlayerBoardsRevealedParser(ParserState, StateFacade),
                    new BattlegroundsOpponentRevealedParser(ParserState, StateFacade),
                    new BattlegroundsHeroSelectedParser(ParserState, StateFacade),
                    new BattlegroundsBattleOverParser(ParserState, StateFacade),
                    new BattlegroundsRerollParser(ParserState, StateFacade),
                    new BattlegroundFreezeParser(ParserState, StateFacade),
                    new BattlegroundsMinionsBoughtParser(ParserState, StateFacade),
                    new BattlegroundsMinionsSoldParser(ParserState, StateFacade),
                    new BattlegroundsHeroKilledParser(ParserState, StateFacade),
                    new BattlegroundsQuestRewardEquippedParser(ParserState, StateFacade),
                    new BattlegroundsQuestRewardDestroyedParser(ParserState, StateFacade),
                    new BloodGemBuffChangedParser(ParserState, StateFacade),
                    new BeetleArmyChangedParser(ParserState, StateFacade),
                    new BallerBuffChangedParser(ParserState, StateFacade),
                    new TotalMagnetizeChangedParser(ParserState, StateFacade),
                    //new BattlegroundsExtraGoldNextTurnParser(ParserState, StateFacade),
                    new BattlegroundsTrinketSelectedParser(ParserState, StateFacade),
                    new BattlegroundsTrinketSelectionParser(ParserState, StateFacade),
                    new BattlegroundsTimewarpedTavernActiveParser(ParserState, StateFacade),

                    new DecklistUpdateParser(ParserState, StateFacade),
                    new GameRunningParser(ParserState, StateFacade),
                    new AttackParser(ParserState, StateFacade),
                    new NumCardsPlayedThisTurnParser(ParserState, StateFacade),
                    new NumCardsDrawnThisTurnParser(ParserState, StateFacade),
                    new HeroPowerUsedParser(ParserState, StateFacade),
                    new GalakrondInvokedParser(ParserState, StateFacade),
                    new CardBuffedInHandParser(ParserState, StateFacade),
                    new MinionGoDormantParser(ParserState, StateFacade),
                    new AttackOnBoardParser(ParserState, StateFacade),
                    new JadeGolemParser(ParserState, StateFacade),
                    new CthunParser(ParserState, StateFacade),
                    new EntityUpdateParser(ParserState, StateFacade),
                    //new ResourcesThisTurnParser(ParserState, StateFacade),
                    new ResourcesUsedThisTurnParser(ParserState, StateFacade),
                    new WhizbangDeckParser(ParserState, StateFacade),
                    new BattlegroundsTavernPrizesParser(ParserState),
                    new LinkedEntityParser(ParserState, StateFacade),
                    new ZoneChangeParser(ParserState, StateFacade),
                    new ZonePositionChangedParser(ParserState, StateFacade),
                    new CostChangedParser(ParserState, StateFacade),
                    new SpawnTimeCountChangedParser(ParserState, StateFacade),
                    new TurnDurationUpdateParser(ParserState, StateFacade),
                    new StartOfGameTriggerParser(ParserState, StateFacade),
                    new DataScriptChangedParser(ParserState, StateFacade),
                    new ImmolateChangedParser(ParserState, StateFacade),
                    new OverloadedCrystalsParser(ParserState, StateFacade),
                    new CorposesSpentThisGameParser(ParserState, StateFacade),
                    new ConstructedAnomalyParser(ParserState, StateFacade),
                    new TouristRevealedParser(ParserState, StateFacade),
                    new ParentCardChangedParser(ParserState, StateFacade),
                    new StarshipLaunchedParser(ParserState, StateFacade),


                    new CardForgedParser(ParserState, StateFacade),
                    new LocationUsedParser(ParserState, StateFacade),
                    new CreateCardInGraveyardParser(ParserState, StateFacade),
                    new MindrenderIlluciaParser(ParserState, StateFacade),
                    new SpecialCardPowerParser(ParserState, StateFacade),
                    new WheelOfDeathCounterParser(ParserState, StateFacade),

                    // Needs to happen after EntityUpdate, because somsetimes the event is sent from the 
                    // SHOW_ENTITY block that triggers the ENTITY_UPDATE event
                    new CopiedFromEntityIdParser(ParserState, StateFacade),
                    new SpecialTargetParser(ParserState, StateFacade),
                };
            }
        }
    }
}
